### PR TITLE
feat(tensor): add mask_select (boolean mask indexing)

### LINKED ISSUE #3669: Boolean mask indexing
### Feature description

Boolean mask indexing, similar to:

```py
inds = scores > self.conf_threshold
boxes, landmarks, scores = boxes[inds], landmarks[inds], scores[inds]
```

From [discord](https://discord.com/channels/1038839012602941528/1038839013735399547/1413497627772719317)

### PR DESCRIPTION
**Pull request by Yashiru**
# PR : masked_select (boolean mask indexing)

**Title:** `feat(tensor): add mask_select (boolean mask indexing)`

**Base:** `tracel-ai:main` ← **Compare:** `Yashiru:feat/masked-select`

---

### Checklist

- [x] Confirmed that `cargo run-checks` command has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

Closes #3669.

### Changes

This implements the approach I proposed in #3669: mirror the existing `nonzero`/`argwhere` design
and keep the PR small. Digging into the codebase refined it in two ways: no new shared helper is
needed at all (the op composes entirely from existing operations), and I ended up diverging from
my own naming proposal to follow a repo convention I discovered along the way (see the naming note
below).

Adds `mask_select`, the equivalent of numpy `tensor[mask]` / `torch.masked_select(tensor, mask)`:
it returns a 1D tensor containing the elements of `self` where the same-shaped `Bool` mask is
`true`, in the order of the flattened input tensor.

Because the number of selected elements is data-dependent, the output shape can't be known without
reading the mask, so this makes the same synchronous-read tradeoff already accepted by `argwhere`
and `nonzero`. A `mask_select_async` variant is provided for non-blocking contexts (e.g. wasm),
mirroring the `nonzero`/`nonzero_async` pattern.

The implementation is a small composition of existing ops, with **no backend trait changes**:

```rust
pub async fn mask_select_async(self, mask: Tensor<D, Bool>) -> Tensor<1, K> {
    check!(TensorCheck::mask_select(&self.shape(), &mask.shape()));
    let indices = mask.flatten::<1>(0, D - 1).argwhere_async().await.squeeze_dim::<1>(1);
    self.flatten::<1>(0, D - 1).select(0, indices)
}
```

Notes:
- Purely additive: no changes to existing APIs, no breaking change.
- Lives in the `K: Basic` block, so it works for `Float`, `Int`, and `Bool` tensors.
- The gradient composes from the existing `reshape`/`select` backwards; no autodiff code is needed
  (covered by autodiff tests).
- `mask` must have the same shape as `self`, enforced by a `TensorCheck` (a mismatched mask would
  otherwise select the wrong elements, or read out of bounds on backends without bounds checks).
  Unlike `torch.masked_select`, the mask is not broadcast; that could be added later if wanted.
- An all-`false` mask yields an empty `[0]` tensor with no special-casing (via `argwhere`'s
  `[0, 1]` shape).
- The docs explicitly flag each call as a host-device synchronization point, with guidance to keep
  it on final results rather than inside performance-critical loops.

**Naming note:** in the issue I proposed `masked_select` for PyTorch parity, but while studying
the codebase I noticed burn deliberately renamed `masked_fill` to `mask_fill` and `torch.where` to
`mask_where`, so I followed the local `mask_*` convention with `mask_select`. Discoverability is
preserved by the book table, which maps it to `tensor.masked_select(mask)` exactly like the
`mask_fill` row. Happy to switch back if you prefer PyTorch-parity naming, it's a quick change.

**Possible follow-ups**, out of scope here: the main cost on GPU backends is inherited from the default
`bool_argwhere` (full mask readback + host loop). A device-side `bool_argwhere` for cubecl (stream
compaction, count-only readback) would speed up `argwhere`, `nonzero`, and `mask_select` at once;
a dedicated fused kernel could come on top of it. I'd be happy to tackle that next if there's
interest.

### Testing

- `tensor/float/ops/mask_select.rs`: 1D, 2D, 3D (flattened order across dims), single-element
  (exercises the `squeeze_dim` path), empty mask (forces execution via `into_data`), transposed /
  flipped / narrowed inputs (non-contiguous views), and a `#[should_panic]` shape-mismatch case.
- `tensor/int/ops/mask_select.rs` and `tensor/bool/ops/mask_select.rs`: `Int` and `Bool` kinds
  (the `Bool` source goes through the dedicated bool select dispatch path).
- `autodiff/mask_select.rs`: gradients flow to the selected positions (zeros elsewhere), plus the
  all-`false` mask backward (zero-size scatter-add).
- The doc example runs as a doctest.
- `cargo run-checks` passes locally.

### CODE REVIEW

**Review by laggui — APPROVED** [review]
LGTM as a first implementation.

We should at least add the tensor ops trait with dispatch even if most backends will currently rely on the argwhere + select default. `burn-tch` can override the default, and other backends can eventually do the same.

Probably worth a separate follow-up issue.

**Comment by Yashiru** [comment]
Thanks for the review and the merge!

Opened #5248 as the follow-up for promoting `mask_select` to a backend tensor op with dispatch,
with the default staying on the `argwhere` + `select` composition and a native `burn-tch` override
to start. I'll follow up with a PR there.
